### PR TITLE
mon, osd: parallel clean_pg_upmaps

### PR DESCRIPTION
**Pull request by xiexingguo**
On Cephalocon Barcelona, I had a talk with Tom Byrne, Senior storage system administrator at the Rutherford Appleton Laboratory (RAL), part of STFC, and author of the impressing "CRUSH-ing the OSD Variance Problem" talk, about his nice, big Ceph cluster that has grown from 1500 OSDs to 5000 OSDs (40PB) during the past few years:

> I am worried about the amount of time to create a new OSD map on our large 5000 OSD, 25000 PG cluster that uses upmap.
> 
> Before we used upmap, our OSD map creation time was 2-3 seconds. After the upmap balancer had balanced the cluster and added ~14000 upmap item entries, the OSD map creation time after any cluster change (stopping OSD, reweighting OSD) took about 15 seconds, which was significantly longer and is causing us issues with the monitors hanging, and blocked requests as the cluster continues to try and talk to the down OSDs.
> 
> Looking at the logs with debugging turned up while the leader monitor generates a new OSD map, I traced the extra OSDmap creation time to the maybe_remove_pg_upmap function. It appears that for any change to the cluster, no matter how small, the maybe_remove_pg_upmap function checks all upmap entries in the OSD map for validity when creating the new OSD map. It seems to do this in a single thread, so the time taken scales with more upmap entries.

With this change, mon should now be able to finish the upmap tidy-up work in approximate 1/8 of the time that the original maybe_remove_pg_upmaps should take, and can be further reduced by switching to use an even larger "**mon_cpu_threads**" value if necessary. 
And below are some results from the raw maybe_remove_pg_upmaps performance testing(5000 OSDs, 10000 pg_upmap_items) on my poor local PC, just FYI:
- **Was: maybe_remove_pg_upmaps (~10000 pg_upmap_items) latency:104s**
- **Now: clean_pg_upmaps (~10000 pg_upmap_items) latency:7s**

Fixes: http://tracker.ceph.com/issues/40104

- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by xiexingguo** [comment]
retest this please

**Comment by xiexingguo** [comment]
retest this please

**Comment by xiexingguo** [comment]
retest this please

**Comment by liewegas** [comment]
My only concern is the shift from a ps range for process() to a vector of pgids for the mapping job.  Would it make sense to make process take a pg_t and have the generic ParallelPGMapper layer work either in terms of ranges or a vector?  That would move the loop out of the process() implementations, which is probably a good thing anyway?

**Comment by xiexingguo** [comment]
The problem is that we'll have to update pending_inc at the end of process, which needs to be done under the protection of a lock and will become super inefficient once processed at the granularity of per pg.
I end up making process accept both kind of inputs, either a set of pgids, or pool + [ps_begin, ps_end),
which I think is a good trade-off between the performance (for the prime pg_temp side) and introducing more code complexity (for clean_pg_upmaps).

**Comment by liewegas** [comment]
> The problem is that we'll have to update pending_inc at the end of process, which needs to be done under the protection of a lock and will become super inefficient once processed at the granularity of per pg.
> I end up making process accept both kind of inputs, either a set of pgids, or pool + [ps_begin, ps_end),
> which I think is a good trade-off between the performance (for the prime pg_temp side) and introducing more code complexity (for clean_pg_upmaps).

Makes sense!

File: src/osd/OSDMap.cc [hunk 1]
```diff
@@ -1855,55 +1881,81 @@ void OSDMap::maybe_remove_pg_upmaps(CephContext *cct,
...
+  ldout(cct, 10) << __func__ << dendl;
+  vector<pg_t> to_check;
+  vector<pg_t> to_cancel;
+  map<pg_t, mempool::osdmap::vector<pair<int,int>>> to_remap;
+
+  get_upmap_pgs(&to_check);
+  for (auto& p : pg_upmap)
+    to_check.push_back(p.first);
+  for (auto& p : pg_upmap_items)
+    to_check.push_back(p.first);
```
Comment (hjwsm1989) [inline src/osd/OSDMap.cc]: the above two for loops job has already done in get_upmap_pgs?

Comment (xiexingguo) [reply src/osd/OSDMap.cc]: That's right, thanks!

Comment (xiexingguo) [reply src/osd/OSDMap.cc]: Just c& p the origin code :man_facepalming:

**Comment by xiexingguo** [comment]
retest this please

**Comment by xiexingguo** [comment]
retest this please

File: src/mon/OSDMonitor.h [hunk 2]
```diff
@@ -249,6 +249,35 @@ class OSDMonitor : public PaxosService {
     FAST_READ_DEFAULT
   };
 
+  struct CleanUpmapJob : public ParallelPGMapper::Job {
+    CephContext *cct;
+    const OSDMap& osdmap;
```
Comment (hjwsm1989) [inline src/mon/OSDMonitor.h]: ParallelPGMapper::Job have a member 'Const OSDMap *osdmap', so we can use that one?

**Comment by xiexingguo** [comment]
http://pulpito.ceph.com/xxg-2019-06-06_02:43:56-rados-wip-40104-distro-basic-smithi/

The only suspicious failure is http://qa-proxy.ceph.com/teuthology/xxg-2019-06-06_02:43:56-rados-wip-40104-distro-basic-smithi/4008172/teuthology.log, which should have been addressed by https://github.com/ceph/ceph/pull/28328

**Comment by xiexingguo** [comment]
@Tom-Byrne Would you mind telling me which version is your Ceph cluster running? For luminous or mimic, I guess I'll need to prepare the backports (which are non-trivial) myself so you can give it a try and see if it really helps.

Thanks!

**Comment by Tom-Byrne** [comment]
@xiexingguo we're currently running Luminous (12.2.12) on the cluster affected by this issue, is it feasible to backport to Luminous?

Thanks for working on this, it's really appreciated!

**Comment by xiexingguo** [comment]
@Tom-Byrne 

> we're currently running Luminous (12.2.12) on the cluster affected by this issue, is it feasible to backport to Luminous?

Sure, see https://github.com/ceph/ceph/pull/28594.

Note that the conflicts are non-trivial and I didn't do any QA.
So you might need to wait 12.2.13 for the fix.

**Comment by Tom-Byrne** [comment]
@xiexingguo Just to let you know, we got this change onto our monitors a little while ago and our OSDMap creation time dropped back down to under 5 seconds, thank you!

**Comment by xiexingguo** [comment]
@Tom-Byrne my pleasure:-)

**Pull request merged**